astropy/cosmology/funcs.py

```diff
@@ -6,35 +6,93 @@
 import warnings
 import numpy as np
 
+try:
+    from scipy.interpolate import CubicSpline
+except ImportError:
+    HAS_SCIPY = False
+else:
+    HAS_SCIPY = True
+
 from .core import CosmologyError
 from astropy.units import Quantity
 
 __all__ = ['z_at_value']
 
 __doctest_requires__ = {'*': ['scipy']}
 
+_interpolation_choices = ['cubic', 'linear']
+
+
+def _z_at_array(func, fvals, zmin, zmax, nbins=10000,
+                logspace=True, interpolation=None):
+    """Helper function to interpolate (func, z) over a grid for array input"""
+    if logspace:
+        zgrid = np.logspace(np.log10(zmin), np.log10(zmax), nbins)
+    else:
+        zgrid = np.linspace(zmin, zmax, nbins)
+
+    fgrid = func(zgrid)
+
+    fvals_val = fvals.value
+    fgrid_val = fgrid.to_value(fvals.unit)
+
+    if interpolation is None:
+        if HAS_SCIPY:
+            interpolation = 'cubic'
+        else:
+            warnings.warn((
+                "SciPy not found, so falling back to linear numpy interpolation"
+                " scheme for array z_at_value, instead of the normal cubic"
+                " spline."))
+            interpolation = 'linear'
+
+    if interpolation not in _interpolation_choices:
+        raise ValueError(
+                f"{interpolation} is an invalid interpolation method."
+                f" Please use an option from {_interpolation_choices}."
+            )
+
+    if interpolation == 'cubic':
+        interpolator = CubicSpline(fgrid_val, zgrid)
```

Comment:
Protect from extrapolation

```suggestion
        interpolator = CubicSpline(fgrid_val, zgrid, extrapolate=False)
```

Reply:
In principle this should not occur anyway, since the grid is set from (zmin, zmax) and results outside of those would be rejected.